### PR TITLE
create FLAVA example using FSDP

### PR DESCRIPTION
**Pull request by edward-io**
Summary:

Test plan:

repro: `torchrun --master_port=1256 --nproc_per_node=8 flava/plain_train.py config=flava/configs/pretraining/debug.yaml`

using slurm: `srun -p $TRAIN_QUEUE --cpus-per-task=96 -t 24:00:00 --gpus-per-node=8 torchrun --master_port=1256 --nproc_per_node=8 flava/plain_train.py config=flava/configs/pretraining/debug.yaml`

check that lightning still works: `traini python -m flava.train config=flava/configs/pretraining/debug.yaml`

Fixes #{issue number}

### CODE REVIEW

File: examples/flava/data/datamodules.py [hunk 1]
```diff
@@ -33,12 +33,20 @@
...
 def transform_image(transform, sample):
     sample.update(transform(sample["image"]))
     return sample
 
+def get_sampler(dataset, shuffle=True):
+    if dist.is_initialized():
+        return DistributedSampler(dataset, shuffle=shuffle)
+    if shuffle:
+        return torch.utils.data.RandomSampler(dataset)
+    return torch.utils.data.SequentialSampler(dataset)
```
Comment (edward-io) [inline examples/flava/data/datamodules.py:49]: this doesn't affect lightning training, but let me know if you want me to revert it @ankitade

File: examples/flava/native/train.py [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+                ) if self.scaler else nullcontext():
+                    output = model(data)
+                    total_loss = self.calculate_loss(output, validation=True)
+                    validation_loss += total_loss.detach()
+
+        print0(validation_loss)
+        dist.reduce(validation_loss, dst=0)
+        norm_validation_loss = validation_loss.item() / dist.get_world_size()
+        print0(f"step {self.steps} EVAL loss: {norm_validation_loss:.4}")
+        self.log("validation/loss", norm_validation_loss, always_log=True)
```
Comment (ankitade) [inline examples/flava/native/train.py:305]: move to compute_loss just like for train?

File: examples/flava/native/utils.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+try:
+    from rich.console import Console
+
+    c = Console(force_terminal=True)
+    print = c.log
+except ImportError:
+    pass
+
+
+def build_config() -> DictConfig:
```
Comment (ankitade) [inline examples/flava/native/utils.py:26]: cant u just use the existing build config method?

File: examples/flava/native/train.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        )
+        strategy = self.config.training.strategy
+
+        print0(
+            f"before {strategy} model parameters: {get_model_parameters(model):,}, "
+            f"size: {get_model_size_gb(model):.3} GB"
+        )
+
+        model = model.to(self.device)
+        print0(f"move model to cuda: {torch.cuda.memory_allocated()/1024**3:.3} GB")
```
Comment (ankitade) [inline examples/flava/native/train.py:123]:
```suggestion
        print0(f"after moving model to cuda: {torch.cuda.memory_allocated()/1024**3:.3} GB")
```

File: examples/flava/native/train.py [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+        while True:
+            t0 = time.time()
+            self.epochs += 1
+            dataloader = self.datamodule.train_dataloader()
+            dataloader.set_epoch(self.epochs)
+
+            for i, data in enumerate(dataloader):
+                torch.cuda.reset_peak_memory_stats()
+
+                self.steps += 1
```
Comment (ankitade) [inline examples/flava/native/train.py:211]: do u need to sync this?

File: torchmultimodal/modules/layers/transformer.py [hunk 6]
```diff
@@ -384,12 +388,22 @@ def forward(
             all_hidden_states.append(hidden_states)
 
             layer_head_mask = head_mask[i] if head_mask is not None else None
-            layer_outputs = layer_module(
-                hidden_states,
-                attention_mask=attention_mask,
-                head_mask=layer_head_mask,
-                return_attn_weights=True,
-            )
+            if self.checkpoint_activations:
```
[the remark this answers is not in the document]
Comment (edward-io) [reply torchmultimodal/modules/layers/transformer.py:391]: hmm how can we get checkpointing then?

File: examples/flava/native/train.py [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+                    "stats/max_gpu_allocated_gb",
+                    torch.cuda.max_memory_allocated() / 1024**3,
+                )
+                # TODO implement imagenet eval
+                # TODO implement checkpoint saving
+
+                self.validate()
+
+    def validate(self):
+        if self.steps % self.config.training.validation_steps != 0 or self.steps == 0:
```
[the remark this answers is not in the document]
Comment (edward-io) [reply examples/flava/native/train.py:279]: skip validation on the first step

File: examples/flava/native/utils.py (hunk 3 above)
Comment (edward-io) [reply examples/flava/native/utils.py:26]: are you referring to the one in examples/flava/utils.py? it has some lightning specific settings

File: examples/flava/native/train.py (hunk 2 above)
Comment (edward-io) [reply examples/flava/native/train.py:305]: calculate_loss doesn't accumulate the loss, so we need this still

File: examples/flava/native/train.py (hunk 5 above)
Comment (edward-io) [reply examples/flava/native/train.py:211]: hm, not that i'm aware. why would they be synced?

File: examples/flava/native/train.py [hunk 8]
```diff
@@ -0,0 +1,349 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+# train using `torchrun --master_port=1256 --nproc_per_node=4 -m flava.native.train config=flava/configs/pretraining/debug.yaml`
```
Comment (ankitade) [inline examples/flava/native/train.py:7]: change the config path

File: examples/flava/native/model.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        **flava_pretraining_kwargs: Any,
+    ):
+        super().__init__()
+        self.model = flava_model_for_pretraining(**flava_pretraining_kwargs)
+        self.use_bf16 = use_bf16
+
+    def forward(self, batch):
+        # super hacky
+        if "encode_text" in batch:
+            return self.model.encode_text(*batch["encode_text"])
```
[the remark this answers is not in the document]
Comment (edward-io) [reply examples/flava/native/model.py:56]: Yes

File: torchmultimodal/modules/layers/transformer.py [hunk 10]
```diff
@@ -387,22 +387,22 @@ def forward(
...
-                    attention_mask=attention_mask,
-                    head_mask=layer_head_mask,
-                    return_attn_weights=True,
-                )
+            layer_outputs = checkpoint(
+                layer_module,
+                hidden_states,
+                attention_mask,
+                layer_head_mask,
+                True,  # return_attn_weights
```
Comment (rohan-varma) [inline torchmultimodal/modules/layers/transformer.py:397]: oof, no `kwarg` support.

File: examples/flava/data/datamodules.py [hunk 11]
```diff
@@ -518,9 +544,12 @@ def test_dataloader(self):
     def _build_dataloader(self, dataset: torch.utils.data.Dataset, shuffle=True):
         return torch.utils.data.DataLoader(
             dataset,
-            shuffle=shuffle,
+            sampler=get_sampler(dataset, shuffle),
```
[the remark this answers is not in the document]
Comment (edward-io) [reply examples/flava/data/datamodules.py]: we need the datamodule for the multidataloader currently, we can remove it in a future PR?

File: examples/flava/native/train.py [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+                check_fn = lambda submodule: isinstance(
+                    submodule, TransformerEncoderLayer
+                )
+                non_reentrant_wrapper = partial(
+                    checkpoint_wrapper,
+                    offload_to_cpu=False,
+                    checkpoint_impl=CheckpointImpl.REENTRANT,
+                )
+                apply_activation_checkpointing_wrapper(
+                    model,
```
Comment (ankitade) [inline examples/flava/native/train.py:200]: remove?

File: examples/flava/native/train.py [hunk 13]
```diff
@@ -0,0 +1,385 @@
...
+import time
+from functools import partial
+from typing import Any, Dict, Tuple, Union
+
+import datasets
+import torch
+import torch.distributed as dist
+from common.data import MultiDataModule
+from flava.definitions import FLAVAArguments
+from flava.native.data import (
```
Comment (ankitade) [inline examples/flava/native/train.py:20]: strangely i am not seeing native/data in this PR

File: examples/flava/native/train.py [hunk 14]
```diff
@@ -0,0 +1,385 @@
...
+        optimizer, scheduler = get_optimizer(
+            model,
+            **self.config.training.optimizer,
+        )
+
+        while True:
+            t0 = time.time()
+            self.epochs += 1
+            dataloader = self.datamodule.train_dataloader()
+            dataloader.set_epoch(self.epochs)
```
Comment (ankitade) [inline examples/flava/native/train.py:257]: wat does this do?

File: examples/flava/native/run.slurm [hunk 15]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
```
[the remark this answers is not in the document]
Comment (ankitade) [reply examples/flava/native/run.slurm:1]: remove

File: examples/flava/native/configs/pretrain_debug.yaml [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+    text_num_attention_heads: 16
+
+    multimodal_num_hidden_layers: 32
+    multimodal_hidden_size: 1280
+    multimodal_intermediate_size: 5120
+    multimodal_num_attention_heads: 16
+
+    codebook_image_size: 8192
+
+  3b:
```
Comment (ankitade) [inline examples/flava/native/configs/pretrain_debug.yaml:123]: i liked the prev config style, one per model. but can do that in a follow up diff

File: examples/flava/native/train.py (hunk 14 above)
Comment (edward-io) [reply examples/flava/native/train.py:257]: sets the RNG seed given an epoch for distributedsampler, see the warning here: https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler

File: examples/flava/native/train.py [hunk 17]
```diff
@@ -0,0 +1,380 @@
...
+            if self.config.training.activation_checkpointing:
+                # note: activation checkpointing wrapper currently is faulty
+                # - non-reentrant does not support kwargs in TransformerEncoderLayer
+                # - memory reduction from checkpointing is less than expected
+
+                check_fn = lambda submodule: isinstance(
+                    submodule, TransformerEncoderLayer
+                )
+                if self.config.training.activation_checkpointing_reentrant:
+                    checkpoint_impl = CheckpointImpl.REENTRANT
```
Comment (ankitade) [inline examples/flava/native/train.py:184]: does this crash?

**Review by ankitade — APPROVED** [review]
